### PR TITLE
ProtoClient: provide process_in function to logwriter

### PR DESCRIPTION
**Pull request by alltilla**
As discussed, it might be better to give clean `process_in()` function from `LogProtoClient` as currently `flush()` does both directions, which complicates the communication between `LogWriter` and `LogProtoClient`.

This cleanup would make it easier to do `ProtoClient` level I/O communications (handshake, result handling, etc...), and it would make cleaner code in `Proto` and `LogWriter` level.

I also implemented a use of this process_in() function. LogProtoTextClient can drop input, if it is set from the config by `close-on-input` option, in case of AFSocket based destinations.
This can help to continue the PR #1433.

Signed-off-by: Attila Szakacs <attila.szakacs@balabit.com>

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by furiel** [comment]
So the problem comes from that collectd sends a reply about the success/failure of the request. But when that happens, syslog-ng closes the connection.

My question is: why do we want to ignore the reply? Me seems it is valuable information that should be used for flow control. Also, it would be useful from debugging point of view: in case of failure, the human readable message part of the reply (that comes after the status number) should be written as an `msg_error`.

I would imagine collectd destination as a threaded destination driver, instead of an afsocket.

**Comment by alltilla** [comment]
Cleaned up the `Writer`/`ProtoClient` side, currently working on `Reader`/`ProtoServer`.
Kindly review the current changes, if you have the time, as it had some changes we did not discuss.
Thanks :smile:

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by furiel** [comment]
@kira-syslogng retest this please

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/logwriter.c [hunk 1]
```diff
@@ -183,12 +184,17 @@ log_writer_set_queue(LogWriter *self, LogQueue *queue)
...
 
   g_assert((self->super.flags & PIF_INITIALIZED) != 0);
-  self->work_result = log_writer_flush(self, LW_FLUSH_NORMAL);
+  if (cond == G_IO_OUT)
+    self->work_result = log_writer_process_out(self);
+  else if (cond == G_IO_IN)
+    self->work_result = log_writer_process_in(self);
+  else
+    self->work_result = FALSE;
 }
```
[the remark this answers is not in the document]
Comment (alltilla) [reply lib/logwriter.c]: Thank you for your finding, fixed it.

File: lib/logwriter.c [hunk 2]
```diff
@@ -261,12 +267,31 @@ log_writer_io_flush_output(gpointer s)
...
+    {
+      main_loop_io_worker_job_submit(&self->io_job, G_IO_IN);
+    }
+  else if (!main_loop_worker_job_quit())
+    {
+      log_writer_work_perform(s, G_IO_IN);
+      log_writer_work_finished(s);
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (alltilla) [reply lib/logwriter.c]: Thanks, agreed. Fixed it.

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/logproto/logproto-client.h [hunk 3]
```diff
@@ -135,6 +136,21 @@ log_proto_client_flush(LogProtoClient *s)
...
+  if (s->process_in)
+    return s->process_in(s);
+  else if (s->flush)
+    /*
+     * In some clients, flush is used for input processing, but it should not be.
+     * Fix these clients, than remove the flush call here.
+     */
+    return s->flush(s);
+  else
+    return LPS_SUCCESS;
```
[the remark this answers is not in the document]
Comment (alltilla) [reply lib/logproto/logproto-client.h]: I was following the `flush()` logic here (see above lines). I can imagine not having `process_in()` defined (and `flush()` defined) acceptable. I would prefer it staying like this, but I am persuadable :D.

File: lib/logwriter.c [hunk 4]
```diff
@@ -315,29 +333,15 @@ log_writer_update_fd_callbacks(LogWriter *self, GIOCondition cond)
...
-        {
-          /* in case the protocol requested G_IO_IN, it means that it needs to
-           * invoke read in the flush code, so just install the flush_output
-           * handler for input */
-
-          iv_fd_set_handler_in(&self->fd_watch, log_writer_io_flush_output);
-        }
+      if (cond & G_IO_IN)
+        iv_fd_set_handler_in(&self->fd_watch, log_writer_io_handle_in);
+      else if (self->flags & LW_DETECT_EOF)
```
[the remark this answers is not in the document]
Comment (alltilla) [reply lib/logwriter.c]: There were more modification here. I removed `&& (cond & G_IO_IN) == 0` from one of the conditions. But I believe, the result is the same.

A little bit of explanation:

Old code:
```
if (self->flags & LW_DETECT_EOF && (cond & G_IO_IN) == 0)
  iv_fd_set_handler_in(&self->fd_watch, log_writer_io_check_eof);
else if (cond & G_IO_IN)
  iv_fd_set_handler_in(&self->fd_watch, log_writer_io_flush_output);
else
  iv_fd_set_handler_in(&self->fd_watch, NULL);
```
This resulted:
`LW_DETECT_EOF == TRUE && G_IO_IN == TRUE`: `flush`
`LW_DETECT_EOF == TRUE && G_IO_IN == FALSE`: `check_eof`
`LW_DETECT_EOF == FALSE && G_IO_IN == TRUE`: `flush`
`LW_DETECT_EOF == FALSE && G_IO_IN == FALSE`: `NULL`

New Code:
```
if (cond & G_IO_IN)
  iv_fd_set_handler_in(&self->fd_watch, log_writer_io_handle_in);
else if (self->flags & LW_DETECT_EOF)
  iv_fd_set_handler_in(&self->fd_watch, log_writer_io_check_eof);
else
  iv_fd_set_handler_in(&self->fd_watch, NULL);
```
This resulted:
`LW_DETECT_EOF == TRUE && G_IO_IN == TRUE`: `handle_in`
`LW_DETECT_EOF == TRUE && G_IO_IN == FALSE`: `check_eof`
`LW_DETECT_EOF == FALSE && G_IO_IN == TRUE`: `handle_in`
`LW_DETECT_EOF == FALSE && G_IO_IN == FALSE`: `NULL`

Which is the same, but sequenced more readable.

But I might have mixed up something, so kindly double check me on this, please :D

File: lib/logproto/logproto-text-client.c [hunk 5]
```diff
@@ -40,6 +40,33 @@ log_proto_text_client_prepare(LogProtoClient *s, gint *fd, GIOCondition *cond, g
   return self->partial != NULL;
 }
 
+static LogProtoStatus
+log_proto_text_client_drop_input(LogProtoClient *s)
+{
+  LogProtoTextClient *self = (LogProtoTextClient *) s;
+  guchar buf[1024] = { 0 };
```
[the remark this answers is not in the document]
Comment (alltilla) [reply lib/logproto/logproto-text-client.c]: Thanks, I will fix this!

File: lib/logproto/logproto-text-client.c [hunk 6]
```diff
@@ -40,6 +40,33 @@ log_proto_text_client_prepare(LogProtoClient *s, gint *fd, GIOCondition *cond, g
...
+  gint rc = -1;
+
+  do
+    {
+      rc = log_transport_read(self->super.transport, buf, sizeof(buf), NULL);
+    }
+  while (rc > 0);
+
+  if (errno != EAGAIN)
```
[the remark this answers is not in the document]
Comment (alltilla) [reply lib/logproto/logproto-text-client.c]: Thanks, good catch. I will fix this.

File: lib/logwriter.c [hunk 7]
```diff
@@ -1263,7 +1243,33 @@ log_writer_flush(LogWriter *self, LogWriterFlushMode flush_mode)
   if (write_error)
     return FALSE;
 
-  return log_writer_flush_finalize(self);
+  LogProtoStatus status = log_proto_client_flush(self->proto);
```
[the remark this answers is not in the document]
Comment (alltilla) [reply lib/logwriter.c]: You are right, I will check this.

File: lib/logproto/logproto-client.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (furiel) [reply lib/logproto/logproto-client.h]: Okok looks good as it is.

File: lib/logwriter.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (furiel) [reply lib/logwriter.c]: Ok thanks for the explanation

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: lib/logwriter.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (alltilla) [reply lib/logwriter.c]: I just brought back the `log_writer_flush_finalize()`, as `log_writer_flush()` returns `gboolean`, which I did not want to change to `LogProtoStatus`.

File: lib/logproto/logproto-text-client.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (alltilla) [reply lib/logproto/logproto-text-client.c]: Fixed.

File: lib/logproto/logproto-text-client.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (alltilla) [reply lib/logproto/logproto-text-client.c]: Fixed, thanks! :)

**Pull request merged**